Chat improvements

dhirallin opened the pull request:
Made some changes to make games and their corresponding chat tabs easier to identify:
- Added the game's ID to the status panel and the title bar for small boards. The ID is removed again when you stop Observing/Examining the game.
- When you hover a game's chat tab, it shows a tooltip with the names of the players 
- Added an info-bar to the top of Game chat tabs, which shows the players and their ratings and also the number of watchers.
- When you click on the game description in the info-bar it maximizes the corresponding game board.
- When you hover the mouse over the number of watchers, it pops up a tooltip that contains the list of all watchers, see screenshot below.
https://www.flickr.com/gp/197116004@N06/FaW6dC7G32

Automatically create and remove chat tabs:
- Now the coresponding chat tab is automatically opened when you Observe/Examine a game, and is removed again when you Unobserve/Unexamine the game. 
- Also opponent chat tabs (which are automatically opened when starting a new game) are now cleaned up when they are no longer needed. Specifically, when you start your next game, any previous private tell chat tabs which are completely empty are closed. Chat tabs containing text are not closed. 

Multi-line* textarea:
- As shown previously, the text-input is now a text-area which expands when the text wraps. The amount of text the user can enter has also been limited to 1023 characters when in the console or when using @ to enter a command and 400 characters when in a chat tab (200 for Guests). This is to correspond with FICS limits, so the user can never enter a message which is too long. 
I was thinking in a future update we might consider allowing <br> tags in tells, then the textarea could be changed to allow true multi-line messages

Chat command truncating and splitting and unicode support:
- Also when using a chat command like tell, whisper etc in the console, the message portion of the command will be truncated to 400 characters and then sent. 
- I added unicode support where tells are first truncated to 400 unicode characters, then  HTML encoded, then the resulting ASCII message is split into multiple 400 character tells and sent. This allows foreign language people to still send 400  unicode characters at once.

Tooltips revamp:
- Added support for "descriptive" tool-tips. These have a title in bold, then a seperator, followed by a longer description. See screenshot below. To create one you add the attribute data-description="some text" to the tooltip triggering element. 

https://flic.kr/p/2pVLeTt

- Changed the way tooltips are placed on mobile etc. I used to have different tooltip placements for mobile and desktop, but I realise this was not the right way to do it. Instead bootstrap supports "fallback placements", where you can specify where to place the tooltip if there is no room for the first choice. So I changed it over to use these.
-  Added support for 'hover-only' tooltips, which don't display when you click a button, but only when you hover over it with a  mouse. These are specified by putting the 'data-tooltip-hover-only' attribute on the button. 
- Found a way to add bootstrap tooltips to the collapsable and dropdown buttons. I made an overlay element with position: absolute which is positioned over the top of the button, and attach the tooltip to it.  

Lastly added a suffix to whispers in chat tabs to differentiate them from kibitzes.

Let me know if you have any issues or thoughts on these. 

Future todo ideas I was pondering:
- Integrate messages into the chat tabs. Basically how this would work, is when the user logs on, a notification appers telling them they have X unread messages and asking if they would like to display them in chat tabs. Then it would create a separate chat tab for each sender. Secondly when the user tries to send a tell to someone who is offline, a dialog would appear asking if they want to send it as a message.
- Add an Adjourn button next to Abort, Resign etc. Perhaps only display this button for standard time controls or slower.
- Add chess piece unicode icons to move list (as discussed)
- Add a text input to the top of the '+' Join Channel button which lets the user type in a person's name and open a chat tab to them. 
- Also add a 'more channels...' button to the Join Channel dropdown which opens a dialog that lets the person subscribe/unsubscribe (+ch/-ch) to more channels. 
- Add a Stats button to the top-right toolbar next to Settings. This would display finger stats and possibly from other commands, except with prettier styling. 
- Add a Tournaments tab to the Play pane, which lets the person easily join upcoming tournaments.
- In settings, have separate settings for board color and theme, so the user can customize the board specifically.

And I want to go back to working on the move list, and making it so the user can eventually annotate and save PGNs properly.

freechessclub-dev commented:
Big PR. Thanks, I love the changes!

> Made some changes to make games and their corresponding chat tabs easier to identify:
> 
> * Added the game's ID to the status panel and the title bar for small boards. The ID is removed again when you stop Observing/Examining the game.
> * When you hover a game's chat tab, it shows a tooltip with the names of the players
> * Added an info-bar to the top of Game chat tabs, which shows the players and their ratings and also the number of watchers.
> * When you click on the game description in the info-bar it maximizes the corresponding game board.
> * When you hover the mouse over the number of watchers, it pops up a tooltip that contains the list of all watchers, see screenshot below.
>   https://www.flickr.com/gp/197116004@N06/FaW6dC7G32

I like the info-bar idea. I had something similar in mind originally for tabs but I couldn't get it right with my limited Bootstrap skills at the time :)

The info-bar could potentially be added to other tabs too -- channel tabs, chat tabs, etc. For instance, channel tab could show the number of users in the channel, chat tab could show if the user was online or offline, or "finger <user> r" on hover, or whatever. I haven't thought it through – but I really like that you made use of that wasted space.

My only concern with this view is that the watchers list is now duplicated. Do you think the original watcher list is still necessary now, or can we remove it and reclaim that space in the game status tab for something else? I'd vote for the latter. The only issue I see is that a user could potentially close their game tab and lose access to the watchers. There's no way to open up the game chat other than, say, whispering or kibitzing. Perhaps there could be a little "chat bubble" icon somewhere to open up game chat. Or we could disallow users from closing a game tab while a game is active (being played, observed or examined). 

> 
> Automatically create and remove chat tabs:
> 
> * Now the coresponding chat tab is automatically opened when you Observe/Examine a game, and is removed again when you Unobserve/Unexamine the game.
> * Also opponent chat tabs (which are automatically opened when starting a new game) are now cleaned up when they are no longer needed. Specifically, when you start your next game, any previous private tell chat tabs which are completely empty are closed. Chat tabs containing text are not closed.

This is nice. I recently added a no-tabs mode because some people coming from legacy interfaces aren't used to tabs and prefer all text in the console. But this setting wasn't personally useful for me. I like tabs, but I am not so fond of the automatic chat tabs. I was considering adding another setting to disable chat tabs but this automatic clean-up of unused tabs is better.

> 
> Multi-line* textarea:
> 
> * As shown previously, the text-input is now a text-area which expands when the text wraps. The amount of text the user can enter has also been limited to 1023 characters when in the console or when using @ to enter a command and 400 characters when in a chat tab (200 for Guests). This is to correspond with FICS limits, so the user can never enter a message which is too long.
>   I was thinking in a future update we might consider allowing  tags in tells, then the textarea could be changed to allow true multi-line messages

Something changed recently for me on mobile and Chrome would no longer autofocus when the input went past the single-line text-input area. Also moving back and forth was starting to be cumbersome especially on mobile. Multi-line input area would really help.

> 
> Chat command truncating and splitting and unicode support:
> 
> * Also when using a chat command like tell, whisper etc in the console, the message portion of the command will be truncated to 400 characters and then sent.
> * I added unicode support where tells are first truncated to 400 unicode characters, then  HTML encoded, then the resulting ASCII message is split into multiple 400 character tells and sent. This allows foreign language people to still send 400  unicode characters at once.

This is so cool! 

> 
> Tooltips revamp:
> 
> * Added support for "descriptive" tool-tips. These have a title in bold, then a seperator, followed by a longer description. See screenshot below. To create one you add the attribute data-description="some text" to the tooltip triggering element.
> 
> https://flic.kr/p/2pVLeTt

Thanks, these are very nice. Descriptive tooltips fill some of the gap of not having any help/documentation. At some point, I was considering adding a "product tour" or walkthrough for new users, i.e. how to seek a game, how to observe a game, etc. , using something like this https://bootstraptour.com. 

> 
> * Changed the way tooltips are placed on mobile etc. I used to have different tooltip placements for mobile and desktop, but I realise this was not the right way to do it. Instead bootstrap supports "fallback placements", where you can specify where to place the tooltip if there is no room for the first choice. So I changed it over to use these.
> * Added support for 'hover-only' tooltips, which don't display when you click a button, but only when you hover over it with a  mouse. These are specified by putting the 'data-tooltip-hover-only' attribute on the button.
> * Found a way to add bootstrap tooltips to the collapsable and dropdown buttons. I made an overlay element with position: absolute which is positioned over the top of the button, and attach the tooltip to it.
> 
> Lastly added a suffix to whispers in chat tabs to differentiate them from kibitzes.
> 
> Let me know if you have any issues or thoughts on these.
> 
> Future todo ideas I was pondering:
> 
> * Integrate messages into the chat tabs. Basically how this would work, is when the user logs on, a notification appers telling them they have X unread messages and asking if they would like to display them in chat tabs. Then it would create a separate chat tab for each sender. Secondly when the user tries to send a tell to someone who is offline, a dialog would appear asking if they want to send it as a message.
> * Add an Adjourn button next to Abort, Resign etc. Perhaps only display this button for standard time controls or slower.
> * Add chess piece unicode icons to move list (as discussed)
> * Add a text input to the top of the '+' Join Channel button which lets the user type in a person's name and open a chat tab to them.
> * Also add a 'more channels...' button to the Join Channel dropdown which opens a dialog that lets the person subscribe/unsubscribe (+ch/-ch) to more channels.
> * Add a Stats button to the top-right toolbar next to Settings. This would display finger stats and possibly from other commands, except with prettier styling.
> * Add a Tournaments tab to the Play pane, which lets the person easily join upcoming tournaments.
> * In settings, have separate settings for board color and theme, so the user can customize the board specifically.
> 
> And I want to go back to working on the move list, and making it so the user can eventually annotate and save PGNs properly.

These all sound good, and fun to implement :) I'd suggest prioritizing the ones related to core game play over chat. This despite the fact that the chat features are personally very useful to me. I think the average user probably cares more about game play over chat, but I could be wrong here. The most requested chat feature has been to be able to assign different colors to different user's tells 🙄

dhirallin commented:
Hold on, just a couple more bugs to fix. Apparently is a catastrophic bottleneck in one of my regex expressions. I thought the 'games' command was acting suspiciously slow lately. But thought it must have been the server lagging, but it was actually the match() function.

dhirallin commented:
The bugs should be fixed now. In regards to watchers. I feel that they're
not inherently a chat thing. And the games chat tab is a little hidden
away, especially on mobile. So I have no problem with the duplication in
itself.

On Mon, 10 June 2024, 4:06 pm freechessclub-dev, ***@***.***>
wrote:

> Big PR. Thanks, I love the changes!
>
> Made some changes to make games and their corresponding chat tabs easier
> to identify:
>
>    - Added the game's ID to the status panel and the title bar for small
>    boards. The ID is removed again when you stop Observing/Examining the game.
>    - When you hover a game's chat tab, it shows a tooltip with the names
>    of the players
>    - Added an info-bar to the top of Game chat tabs, which shows the
>    players and their ratings and also the number of watchers.
>    - When you click on the game description in the info-bar it maximizes
>    the corresponding game board.
>    - When you hover the mouse over the number of watchers, it pops up a
>    tooltip that contains the list of all watchers, see screenshot below.
>    ***@***.***/FaW6dC7G32
>
> I like the info-bar idea. I had something similar in mind originally for
> tabs but I couldn't get it right with my limited Bootstrap skills at the
> time :)
>
> The info-bar could potentially be added to other tabs too -- channel tabs,
> chat tabs, etc. For instance, channel tab could show the number of users in
> the channel, chat tab could show if the user was online or offline, or
> "finger r" on hover, or whatever. I haven't thought it through – but I
> really like that you made use of that wasted space.
>
> My only concern with this view is that the watchers list is now
> duplicated. Do you think the original watcher list is still necessary now,
> or can we remove it and reclaim that space in the game status tab for
> something else? I'd vote for the latter. The only issue I see is that a
> user could potentially close their game tab and lose access to the
> watchers. There's no way to open up the game chat other than, say,
> whispering or kibitzing. Perhaps there could be a little "chat bubble" icon
> somewhere to open up game chat. Or we could disallow users from closing a
> game tab while a game is active (being played, observed or examined).
>
> Automatically create and remove chat tabs:
>
>    - Now the coresponding chat tab is automatically opened when you
>    Observe/Examine a game, and is removed again when you Unobserve/Unexamine
>    the game.
>    - Also opponent chat tabs (which are automatically opened when
>    starting a new game) are now cleaned up when they are no longer needed.
>    Specifically, when you start your next game, any previous private tell chat
>    tabs which are completely empty are closed. Chat tabs containing text are
>    not closed.
>
> This is nice. I recently added a no-tabs mode because some people coming
> from legacy interfaces aren't used to tabs and prefer all text in the
> console. But this setting wasn't personally useful for me. I like tabs, but
> I am not so fond of the automatic chat tabs. I was considering adding
> another setting to disable chat tabs but this automatic clean-up of unused
> tabs is better.
>
> Multi-line* textarea:
>
>    - As shown previously, the text-input is now a text-area which expands
>    when the text wraps. The amount of text the user can enter has also been
>    limited to 1023 characters when in the console or when using @ to enter a
>    command and 400 characters when in a chat tab (200 for Guests). This is to
>    correspond with FICS limits, so the user can never enter a message which is
>    too long.
>    I was thinking in a future update we might consider allowing tags in
>    tells, then the textarea could be changed to allow true multi-line messages
>
> Something changed recently for me on mobile and Chrome would no longer
> autofocus when the input went past the single-line text-input area. Also
> moving back and forth was starting to be cumbersome especially on mobile.
> Multi-line input area would really help.
>
> Chat command truncating and splitting and unicode support:
>
>    - Also when using a chat command like tell, whisper etc in the
>    console, the message portion of the command will be truncated to 400
>    characters and then sent.
>    - I added unicode support where tells are first truncated to 400
>    unicode characters, then HTML encoded, then the resulting ASCII message is
>    split into multiple 400 character tells and sent. This allows foreign
>    language people to still send 400 unicode characters at once.
>
> This is so cool!
>
> Tooltips revamp:
>
>    - Added support for "descriptive" tool-tips. These have a title in
>    bold, then a seperator, followed by a longer description. See screenshot
>    below. To create one you add the attribute data-description="some text" to
>    the tooltip triggering element.
>
> https://flic.kr/p/2pVLeTt
>
> Thanks, these are very nice. Descriptive tooltips fill some of the gap of
> not having any help/documentation. At some point, I was considering adding
> a "product tour" or walkthrough for new users, i.e. how to seek a game, how
> to observe a game, etc. , using something like this
> https://bootstraptour.com.
>
>
>    - Changed the way tooltips are placed on mobile etc. I used to have
>    different tooltip placements for mobile and desktop, but I realise this was
>    not the right way to do it. Instead bootstrap supports "fallback
>    placements", where you can specify where to place the tooltip if there is
>    no room for the first choice. So I changed it over to use these.
>    - Added support for 'hover-only' tooltips, which don't display when
>    you click a button, but only when you hover over it with a mouse. These are
>    specified by putting the 'data-tooltip-hover-only' attribute on the button.
>    - Found a way to add bootstrap tooltips to the collapsable and
>    dropdown buttons. I made an overlay element with position: absolute which
>    is positioned over the top of the button, and attach the tooltip to it.
>
> Lastly added a suffix to whispers in chat tabs to differentiate them from
> kibitzes.
>
> Let me know if you have any issues or thoughts on these.
>
> Future todo ideas I was pondering:
>
>    - Integrate messages into the chat tabs. Basically how this would
>    work, is when the user logs on, a notification appers telling them they
>    have X unread messages and asking if they would like to display them in
>    chat tabs. Then it would create a separate chat tab for each sender.
>    Secondly when the user tries to send a tell to someone who is offline, a
>    dialog would appear asking if they want to send it as a message.
>    - Add an Adjourn button next to Abort, Resign etc. Perhaps only
>    display this button for standard time controls or slower.
>    - Add chess piece unicode icons to move list (as discussed)
>    - Add a text input to the top of the '+' Join Channel button which
>    lets the user type in a person's name and open a chat tab to them.
>    - Also add a 'more channels...' button to the Join Channel dropdown
>    which opens a dialog that lets the person subscribe/unsubscribe (+ch/-ch)
>    to more channels.
>    - Add a Stats button to the top-right toolbar next to Settings. This
>    would display finger stats and possibly from other commands, except with
>    prettier styling.
>    - Add a Tournaments tab to the Play pane, which lets the person easily
>    join upcoming tournaments.
>    - In settings, have separate settings for board color and theme, so
>    the user can customize the board specifically.
>
> And I want to go back to working on the move list, and making it so the
> user can eventually annotate and save PGNs properly.
>
> These all sound good, and fun to implement :) I'd suggest prioritizing the
> ones related to core game play over chat. This despite the fact that the
> chat features are personally very useful to me. I think the average user
> probably cares more about game play over chat, but I could be wrong here.
> The most requested chat feature has been to be able to assign different
> colors to different user's tells 🙄
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/freechessclub/freechessclub-app/pull/102#issuecomment-2157339014>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJEX5DZKRLFZ736C3YEVVDZGU67TAVCNFSM6AAAAABI43UNFSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNJXGMZTSMBRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
